src/Components/Analyzers/src/ComponentInternalUsageDiagnosticAnalzyer.cs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.AspNetCore.Components.Analyzers;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.Extensions.Internal
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class ComponentInternalUsageDiagnosticAnalyzer : DiagnosticAnalyzer
+    {
+        private readonly InternalUsageAnalyzer _inner;
+
+        public ComponentInternalUsageDiagnosticAnalyzer()
+        {
+            // We don't have in *internal* attribute in Blazor.
+            _inner = new InternalUsageAnalyzer(IsInInternalNamespace, hasInternalAttribute: null, DiagnosticDescriptors.DoNotUseRenderTreeTypes);
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DiagnosticDescriptors.DoNotUseRenderTreeTypes);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            _inner.Register(context);
+        }
+
+        private static bool IsInInternalNamespace(ISymbol symbol)
+        {
+            if (symbol?.ContainingNamespace?.ToDisplayString() is string ns)
+            {
+                return string.Equals(ns, "Microsoft.AspNetCore.Components.RenderTree");
```

Comment:
I'm unsure whether we specifically stated which types to consider pubternal. I was expecting this to include `RenderTreeFrame`, `RenderTreeEdit`, `RenderTreeDiff`, or anything else that you would only use when building a custom renderer.

I *didn't* expect us to be forbidding use of `RenderTreeBuilder`, which people would need to do if implementing `IComponent` directly, or writing custom builder logic which I know many component vendors are doing. Should we add a specific exclusion for RTB or maybe move it to a different namespace?

(You previously suggested renaming the `.RenderTree` namespace to `.Rendering` which sounds good to me)

Reply:
As previously discussed, the plan to to move `RenderTreeBuilder` to another namespace.

Reply:
Ok good.